src/tendencies/hyperdiffusion.jl

```diff
@@ -93,32 +118,57 @@ function hyperdiffusion_tendency!(Yₜ, Y, p, t)
                 Spaces.weighted_dss_ghost!,
             )
+                # DSS on Grid scale quantities
+                # Need to split the DSS computation here, because our DSS 
+                # operations do not accept Covariant123Vector types
+                @. ᶜ∇²uₕ = C12(ᶜ∇²u)
+                @. ᶜ∇²uᵥ = C3(ᶜ∇²u)
                 dss_op!(ᶜ∇²uₕ, buffer.ᶜ∇²uₕ)
+                dss_op!(ᶜ∇²uᵥ, buffer.ᶜ∇²uᵥ)
+                @. ᶜ∇²u = C123(ᶜ∇²uₕ) + C123(ᶜ∇²uᵥ)
                 dss_op!(ᶜ∇²specific_energy, buffer.ᶜ∇²specific_energy)
+
+                # DSS on Sub-grid scale quantities
                 if n > 0
                     dss_op!(ᶜ∇²tke⁰, buffer.ᶜ∇²tke⁰)
-                    dss_op!(ᶠ∇²wʲs, buffer.ᶠ∇²wʲs)
+                    # Need to split the DSS computation here, because our DSS 
+                    # operations do not accept Covariant123Vector types         
+                    @. ᶜ∇²uₕʲs = C12(ᶜ∇²uʲs)
+                    @. ᶜ∇²uᵥʲs = C3(ᶜ∇²uʲs)
+                    dss_op!(ᶜ∇²uₕʲs, buffer.ᶜ∇²uₕʲs)
+                    dss_op!(ᶜ∇²uᵥʲs, buffer.ᶜ∇²uᵥʲs)
+                    @. ᶜ∇²uʲs = C123(ᶜ∇²uₕʲs) + C123(ᶜ∇²uᵥʲs)
                     dss_op!(ᶜ∇²specific_energyʲs, buffer.ᶜ∇²specific_energyʲs)
                 end
             end
         end
     end
 
-    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²uₕ)))
-    # Without the C12(), the right-hand side would be a C1 or C2 in 2D space.
+    # Grid scale hyperdiffusion continued
+    @. Yₜ.c.uₕ -= κ₄ * divergence_damping_factor * C12(wgradₕ(divₕ(ᶜ∇²u)))
+    # Without the C123(), the right-hand side would be a C1 or C2 in 2D space.
     if point_type <: Geometry.Abstract3DPoint
-        # TODO: Are these vector conversions correct when there is topography?
-        @. Yₜ.c.uₕ += κ₄ * C12(wcurlₕ(C3(curlₕ(ᶜ∇²uₕ))))
+        ᶜχ₁₂₃ = @. C123(curlₕ(C12(ᶜ∇²u))) + C123(curlₕ(C3(ᶜ∇²u)))
+        @. Yₜ.c.uₕ += κ₄ * (C12(wcurlₕ(C12(ᶜχ₁₂₃))) + C12(wcurlₕ(C3(ᶜχ₁₂₃))))
+        # Reuse the buffer variable here so we don't need an extra one
+        @. ᶜ∇²uᵥ = C3(wcurlₕ(C12(ᶜχ₁₂₃))) + C3(wcurlₕ(C3(ᶜχ₁₂₃)))
+        @. Yₜ.f.u₃ += κ₄ * ᶠwinterp(ᶜJ * ᶜρ, ᶜ∇²uᵥ)
     end
-
     ᶜρ_energyₜ = :θ in propertynames(ᶜspecific) ? Yₜ.c.ρθ : Yₜ.c.ρe_tot
     @. ᶜρ_energyₜ -= κ₄ * wdivₕ(Y.c.ρ * gradₕ(ᶜ∇²specific_energy))
+
+    # Sub-grid scale hyperdiffusion continued
     if n > 0
         @. Yₜ.c.sgs⁰.ρatke -= κ₄ * wdivₕ(ᶜρa⁰ * gradₕ(ᶜ∇²tke⁰))
     end
     for j in 1:n
-        @. Yₜ.f.sgsʲs.:($$j).u₃.components.data.:1 -=
-            κ₄ * wdivₕ(gradₕ(ᶠ∇²wʲs.:($$j)))
+        if point_type <: Geometry.Abstract3DPoint
+            ᶜχ₁₂₃ = @. C123(curlₕ(C12(ᶜ∇²uʲs.:($$j)))) +
+               C123(curlₕ(C3(ᶜ∇²uʲs.:($$j))))
+            # Reuse the buffer variable here so we don't need an extra one
+            @. ᶜ∇²uᵥʲs.:($$j) = C3(wcurlₕ(C12(ᶜχ₁₂₃))) + C3(wcurlₕ(C3(ᶜχ₁₂₃)))
+            @. Yₜ.f.sgsʲs.:($$j).u₃ += κ₄ * ᶠwinterp(ᶜJ * ᶜρ, ᶜ∇²uᵥʲs.:($$j))
```

Comment:
I'll double check with Tapio if we want ᶜρ or ᶜρʲ here, but you don't need to wait for this to merge this PR.

Reply:
We can just use what you have now.

Reply:
Thanks @szy21 , there are still some allocation issues which we will work through before another round of ci checks.